### PR TITLE
Tri state

### PR DESCRIPTION
**Pull request by CLopMan**
Tri-state implementation with parametrized signal size. 

This pr also groups basic components (i.e. reg.vhdl, generic_multiplexer.vhdl, tristate.vhdl) in one folder inside src.

### CODE REVIEW

File: src/BasicComponents/triestate.vhdl [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+architecture Rtl of TriState is
+begin
+    process(activate)
+    begin
+        if activate = '1' then
+            data_out <= data_in;
+        else
+            data_out <= (others => 'Z');
+        end if;
+    end process;
```
Comment (ALVAROPING1) [inline src/BasicComponents/triestate.vhdl]: This process needs to have `data_in` in the sensitivity list, since otherwise it only updates when the `activate` signal changes. It could also be more compactly written with a concurrent statement:
```suggestion
    data_out <= data_in         when activate = '1' else
                (others => 'Z') when activate = '0' else
                (others => 'X');
```

File: tests/tristate_tb.vhdl [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+    T2: entity Src.Tristate
+        generic map(
+            data_size => SIZE
+        )
+        port map(
+            activate => activateT2,
+            data_in => st2,
+            data_out => bus8b
+        );
```
Comment (ALVAROPING1) [inline tests/tristate_tb.vhdl:70]: This should have an extra level of indentation
```suggestion
    T8: entity Src.Tristate
            generic map(
                data_size => SIZE
            )
            port map(
                activate => s_activate,
                data_in => s_data_in8,
                data_out => s_data_out8
            );

    T16: entity Src.Tristate
            generic map(
                data_size => SIZE2
            )
            port map(
                activate => s_activate,
                data_in => s_data_in16,
                data_out => s_data_out16
            );

    -- various to one bus
    T1: entity Src.Tristate
            generic map(
                data_size => SIZE
            )
            port map(
                activate => activateT1,
                data_in => st1,
                data_out => bus8b
            );

    T2: entity Src.Tristate
            generic map(
                data_size => SIZE
            )
            port map(
                activate => activateT2,
                data_in => st2,
                data_out => bus8b
            );
```

File: tests/tristate_tb.vhdl [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+        for i in indiv_tests'range loop
+            report "test: " & integer'image(i + 1);
+            -- inputs
+            s_data_in8 <= indiv_tests(i).data_in8;
+            s_data_in16 <= indiv_tests(i).data_in16;
+            s_activate <= indiv_tests(i).activate;
+
+            wait for 10 ns;
+            -- output
+            --report "debug::: " &  to_string(internal);
```
Comment (ALVAROPING1) [inline tests/tristate_tb.vhdl]: the `internal` signal isn't defined. Did you forget to delete this commented line when copy/pasting?

File: tests/tristate_tb.vhdl [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        end record;
+
+        type test_case2 is record
+            -- inputs
+            activatet1: std_logic;
+            activatet2: std_logic;
+            -- output
+            expected_bus_data: std_logic_vector(SIZE - 1 downto 0);
+        end record;
+
```
Comment (ALVAROPING1) [inline tests/tristate_tb.vhdl]: line with whitespaces at the end
```suggestion

```

File: tests/tristate_tb.vhdl [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+                -- in
+                '1',
+                std_logic_vector(to_unsigned(5, SIZE)),
+                std_logic_vector(to_unsigned(10, SIZE2)),
+                -- out
+                std_logic_vector(to_unsigned(5, SIZE)),
+                std_logic_vector(to_unsigned(10, SIZE2))
+            )
+        );
+
```
Comment (ALVAROPING1) [inline tests/tristate_tb.vhdl]: line with whitespaces at the end
```suggestion

```

File: tests/tristate_tb.vhdl [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+        activatet1 <= bus_tests(i).activatet1;
+        activatet2 <= bus_tests(i).activatet2;
+        wait for 10 ns;
+        -- output
+        assert bus8b = bus_tests(i).expected_bus_data
+            report "[2T1B] failed test " & integer'image(i + 1)
+            & " expected: " & to_string(st1)
+            & " real: " & to_string(bus8b);
+
+        end loop;
```
Comment (ALVAROPING1) [inline tests/tristate_tb.vhdl]: this should be properly indented
```suggestion
        for i in bus_tests'range loop
            -- inputs
            activatet1 <= bus_tests(i).activatet1;
            activatet2 <= bus_tests(i).activatet2;
            wait for 10 ns;
            -- output
            assert bus8b = bus_tests(i).expected_bus_data
                report "[2T1B] failed test " & integer'image(i + 1)
                & " expected: " & to_string(st1)
                & " real: " & to_string(bus8b);
        end loop;
```

File: tests/tristate_tb.vhdl [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+            st1
+
+        ),
+        -- allow t2
+        (
+            -- input
+            '0', '1',
+            -- output
+            st2
+        )
```
Comment (ALVAROPING1) [inline tests/tristate_tb.vhdl:150]: Should we add another 2 tests to check for bus conflicts (both triestates try to write in it) and empty bus (neither triestate writes in it)?

File: tests/tristate_tb.vhdl (hunk 2 above)
Comment (CLopMan) [reply tests/tristate_tb.vhdl:70]: don't think so.  That extra indentation level adds up to three, while the previous line has only one indentation level.

File: src/BasicComponents/triestate.vhdl (hunk 1 above)
Comment (CLopMan) [reply src/BasicComponents/triestate.vhdl]: Does the concurrent statement have any advantages? Otherwise I personally prefer the if statement.

Add the extra case with 'x' value is a good idea, though.

Comment (ALVAROPING1) [reply src/BasicComponents/triestate.vhdl]: the only advantage is shorter code and not having to worry about missing signals in the sensitivity list

File: tests/tristate_tb.vhdl (hunk 7 above)
Comment (CLopMan) [reply tests/tristate_tb.vhdl:150]: Tri-states exists to prevent the first situation from happening. That case does not validate the component functioning, in my opinion. There's maybe something I am not thinking about? Any way, is there a predictable result (as an expected output) in that case?

The second case is a consequence of the current tests. So it is not necessary.

File: tests/tristate_tb.vhdl (hunk 2 above)
Comment (ALVAROPING1) [reply tests/tristate_tb.vhdl:70]: yeah, my mistake. I got confused because label made the entity and generic names align vertically

File: tests/tristate_tb.vhdl (hunk 7 above)
Comment (ALVAROPING1) [reply tests/tristate_tb.vhdl:150]: When there is a conflict in an `std_logic` signal, the value is set to `'X'` to indicate it. My idea was to also check for that so we can be sure that's the result of an invalid input and use it for debugging later on

File: tests/tristate_tb.vhdl (hunk 3 above)
Comment (CLopMan) [reply tests/tristate_tb.vhdl]: probably, thanks for make me noticing.

File: tests/tristate_tb.vhdl (hunk 7 above)
Comment (CLopMan) [reply tests/tristate_tb.vhdl:150]: In that case, i can check it but i  think that shouldn't remain as a test for a component.

**Review by ALVAROPING1 — APPROVED** [review]
LGTM
